BLD: mingw-w64 build fixes

lazka opened the pull request:
Two commits to make the build with mingw-w64 work. Note that this is using mingw-w64 Python, which is not officially supported by upstream CPython. I hope the changes here are small/clear enough to still be considered.

* BLD: Skip MSVC compatibility when building against mingw-w64 Python
* BLD: Fix mingw detection when building with Clang

See the commit messages for details

lazka committed:
BLD: Skip MSVC compatibility when building against mingw-w64 Python

In MSYS2 we build numpy like Linux distros, so we prefer dynamic
linking and it needs to be ABI compatible with all other packages.
numpy by default statically links ucrt and changes the ABI to match
MSVC, which makes it incompatible with other packages.

To differentiate the two use cases detect if we are building against
a mingww-64 Python build, and if we are skip all the special adjustments.

Note that mingw-w64 is not officially supported by upstream CPython.

lazka committed:
BLD: Fix mingw detection when building with Clang

mingw-w64 nowadays is also available in combination with Clang,
for example https://github.com/mstorsjo/llvm-mingw or MSYS2
provide Clang builds using mingw-w64.

Instead of testing for GCC check that __MINGW32__ is defined.

charris commented:
Hmm, seems reasonable to me.

rgommers commented on numpy/_core/meson.build:
```diff
@@ -281,7 +281,7 @@ else
   # but that is broken because we don't have the exact long double
   # representation checks.
-  if cc.get_argument_syntax() != 'msvc'
+  if not is_windows
```
Hmm, this seems fine but really I should fix the feature check. Will try that.

lazka replied:
yeah, note that meson is using set10(), while the code just checks via"ifdef", which is doubly confusing :) I tried to not touch it.

rgommers replied:
This check should be fixed by gh-24979. It would be useful if you could double check that that works for Mingw too.

lazka replied:
I've dropped this patch, now that it is no longer needed.

rgommers commented on numpy/meson.build:
```diff
@@ -10,15 +10,20 @@ endif
 # Platform detection
 is_windows = host_machine.system() == 'windows'
-is_mingw = is_windows and cc.get_id() == 'gcc'
+is_mingw = is_windows and cc.get_id() in ['gcc', 'clang']
```
This can't be right I think, it will break using regular Clang with MSVC-built CPython. Conda-forge uses that for SciPy since recently, so it's not a "no one will do that" situation.

This check is for "this is a Mingw" compiler, so I don't quite understand how you'd get Clang here?

lazka replied:
Do you mean `clang-cl`? `clang` on Windows means clang providing a gcc interface, I only know mingw providing that: https://mesonbuild.com/Reference-tables.html#compiler-ids

rgommers replied:
@h-vetinari conda-forge does have a `clang.exe` with GCC syntax in addition to the main `clang-cl.exe`, right?

h-vetinari replied:
Yes, we have both

lazka replied:
And does clang.exe with GCC syntax use mingw-w64?

rgommers replied:
I had a look for the answer in conda-forge feedstocks, but could find it so quickly. https://github.com/conda-forge/clang-compiler-activation-feedstock/blob/main/recipe/meta.yaml#L19 skips Windows.

h-vetinari replied:
The recipe for clang on windows is in a different [place](https://github.com/conda-forge/clang-win-activation-feedstock/).

> And does clang.exe with GCC syntax use mingw-w64?

No, mingw isn't necessary to run clang.

lazka replied:
I see. I just installed clang from upstream and played around with meson, and it indeed is as you said. I for some reason never encountered this combination. I'll guess I'll best check for `__MINGW32__ ` then. Thanks!

lazka replied:
I've added a check for `__MINGW32__` instead.

mattip replied:
What does cc.get_id() show for mingw? 

FWIW: the intel oneAPI compiler is also a clang-cl based compiler It returns a intel specific string for `get_id()`: `intel-llvm-cl`

lazka replied:
for gcc it returns 'gcc' and clang it returns 'clang'

mattip replied:
> I've added a check for `__MINGW32__` instead.

Then I guess this is the best that can be done. Is there a `__MINGW64__` as well?

lazka replied:
> __MINGW64__

yes, that's defined for 64 only, while `__MINGW32__` is defined for 32 and 64 bit, same as `_WIN32` and `_WIN64`

rgommers commented on numpy/meson.build:
```diff
@@ -10,15 +10,20 @@ endif
+is_mingw = is_windows and cc.get_id() in ['gcc', 'clang']
+is_mingw_python = is_mingw and run_command(
+  py, ['-c', 'import sysconfig; print(sysconfig.get_platform())'],
+  check: true).stdout().strip().startswith('mingw')
```
This variable is only used within the next block, so can you move it into `if `is_mingw`? That avoids running it when we already know that we're not using Mingw.

rgommers replied:
Also, could you change the name to `is_mingw_built_python`, to make it a little clearer?

lazka replied:
It's already guarded by `is_mingw`, but sure, sounds good.

lazka replied:
I've renamed the variable and moved it into the suggested block.

rgommers reviewed (commented):
Thanks @lazka. Yes, we're happy to support Mingw-built Python, that seems like a worthwhile thing and I never really understand the problem CPython devs had with accepting patches.

A few comments on these changes.

lazka commented:
* I've rebased to get #24979 and removed my patch
* I've fixed the check for clang to check for `__MINGW32__` instead of just clang with a gnu interface
* I've renamed `is_mingw_python` to `is_mingw_built_python` and moved it into the block guarded by `is_mingw`

mattip commented:
Thanks @lazka

lazka commented:
Thanks!

```console
user@desktop CLANG64 ~
$ python3 -m venv _venvnumpy
user@desktop CLANG64 ~
$ source _venvnumpy/bin/activate
(_venvnumpy)
user@desktop CLANG64 ~
$ pip install git+https://github.com/numpy/numpy
Collecting git+https://github.com/numpy/numpy
  Cloning https://github.com/numpy/numpy to c:/msys64/tmp/pip-req-build-fnyfgf0b
  Running command git clone --filter=blob:none --quiet https://github.com/numpy/numpy C:/msys64/tmp/pip-req-build-fnyfgf0b
  Resolved https://github.com/numpy/numpy to commit 29cbb1f0a7f339dc350ecf3fc81de9e9c570f6db
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: numpy
  Building wheel for numpy (pyproject.toml) ... done
  Created wheel for numpy: filename=numpy-2.0.0.dev0-cp311-cp311-mingw_x86_64_clang.whl size=7316554 sha256=d993e6729218e1aecb5fcfe84a33e999456a7dc53c73606780e2e2ee98ddb0a2
  Stored in directory: C:/msys64/tmp/pip-ephem-wheel-cache-n4_gvqoz/wheels/61/82/88/254b39bf4fea16d2cf792805b8284966388b27bdfe9533da7c
Successfully built numpy
Installing collected packages: numpy
Successfully installed numpy-2.0.0.dev0

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: python3.exe -m pip install --upgrade pip
(_venvnumpy)
user@desktop CLANG64 ~
$ python3 -c "import numpy; print(numpy)"
<frozen importlib._bootstrap>:241: Warning: Numpy built with MINGW-W64 on Windows 64 bits is experimental, and only available for
testing. You are advised not to use it for production.

CRASHES ARE TO BE EXPECTED - PLEASE REPORT THEM TO NUMPY DEVELOPERS
<module 'numpy' from 'C:/msys64/home/user/_venvnumpy/lib/python3.11/site-packages/numpy/__init__.py'>
(_venvnumpy)
user@desktop CLANG64 ~
```
